### PR TITLE
Refactor ImportanceEvaluators

### PR DESCRIPTION
**Pull request by contramundum53**
## Motivation
There has been duplicates in preprocessing code of MeanDecreaseImpurityImportanceEvaluator and FanovaImportanceEvaluator. We propose to refactor the code and remove the duplication.

This is a remake of #3552.

## Description of the changes

* Refactor `BaseImportanceEvaluator` and its implementations.
* Add missing tests for `ShapleyImportanceEvaluator`.
* Fix `ShapleyImportanceEvaluator` to support categorical variables.

### CODE REVIEW

File: optuna/importance/_fanova/_evaluator.py [hunk 1]
```diff
@@ -79,67 +94,45 @@ def evaluate(
...
-            trials.append(trial)
-
-        trans = _SearchSpaceTransform(distributions, transform_log=False, transform_step=False)
-
-        n_trials = len(trials)
-        trans_params = numpy.empty((n_trials, trans.bounds.shape[0]), dtype=numpy.float64)
-        trans_values = numpy.empty(n_trials, dtype=numpy.float64)
+        non_single_distributions, single_distributions = _split_nonsingle_and_single_distributions(
+            distributions
+        )
```
Comment (c-bata) [inline optuna/importance/_fanova/_evaluator.py:107]: For me, previous implementations looks simpler than your changes.`_split_nonsingle_and_single_distributions` function is not informative very much. What do you think?

```suggestion
        single_distributions = {name: 0.0 for name, dist in distributions.items() if dist.single()}
        non_single_distributions = {name: dist for name, dist in distributions.items() if not dist.single()}
```

File: optuna/importance/_fanova/_evaluator.py [hunk 2]
```diff
@@ -3,17 +3,32 @@
...
+
+
+def _split_nonsingle_and_single_distributions(
+    distributions: Dict[str, BaseDistribution]
+) -> Tuple[Dict[str, BaseDistribution], Dict[str, BaseDistribution]]:
+    non_single_distributions = {
+        name: dist for name, dist in distributions.items() if not dist.single()
+    }
+    single_distributions = {name: dist for name, dist in distributions.items() if dist.single()}
+    return (non_single_distributions, single_distributions)
```
Comment (c-bata) [inline optuna/importance/_fanova/_evaluator.py:31]: Please see the suggestoin at L105.

```suggestion
```

File: optuna/importance/_fanova/_evaluator.py [hunk 3]
```diff
@@ -79,67 +94,45 @@ def evaluate(
...
+            non_single_distributions, transform_log=False, transform_step=False
+        )
 
-        # Many (deep) copies of the search spaces are required during the tree traversal and using
-        # Optuna distributions will create a bottleneck.
-        # Therefore, search spaces (parameter distributions) are represented by a single
-        # `numpy.ndarray`, coupled with a list of flags that indicate whether they are categorical
-        # or not.
+        trans_params: numpy.ndarray = _get_trans_params(trials, trans)
+        values: numpy.ndarray = _get_target_values(trials, target)
```
Comment (c-bata) [inline optuna/importance/_fanova/_evaluator.py:119]: [nits] I think target_values is more informative.

```suggestion
        target_values: numpy.ndarray = _get_target_values(trials, target)
```

Comment (c-bata) [reply optuna/importance/_fanova/_evaluator.py:119]: Or we may pass it to `evaluator.fit()` directly like.

```python
        evaluator.fit(
            X=_get_trans_params(trials, trans),
            y=_get_target_values(trials, target),
            search_spaces=trans.bounds,
            column_to_encoded_columns=trans.column_to_encoded_columns,
        )
```

What do you think?

File: optuna/importance/_fanova/_evaluator.py [hunk 4]
```diff
@@ -79,67 +94,45 @@ def evaluate(
...
-            search_spaces=trans_bounds,
-            column_to_encoded_columns=column_to_encoded_columns,
+            y=values,
+            search_spaces=trans.bounds,
+            column_to_encoded_columns=trans.column_to_encoded_columns,
         )
+        param_importances = numpy.array(
+            [evaluator.get_importance(i)[0] for i in range(len(non_single_distributions))]
+        )
+        param_importances /= numpy.sum(param_importances)
```
Comment (c-bata) [inline optuna/importance/_fanova/_evaluator.py:131]: Looks clear 👍

File: optuna/importance/_fanova/_evaluator.py (hunk 3 above)
Comment (contramundum53) [reply optuna/importance/_fanova/_evaluator.py]: Thanks, I renamed `values` to `target_values`. PTAL. @c-bata

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Basically LGTM. I've left one suggestion on the comment style by following the coding style.

File: tests/integration_tests/test_shap.py [hunk 5]
```diff
@@ -1,20 +1,30 @@
...
-    x1 = trial.suggest_float("x1", 0.1, 3)
-    x2 = trial.suggest_float("x2", 0.1, 3, log=True)
-    x3 = trial.suggest_float("x3", 2, 4, log=True)
-    return x1 + x2 * x3
+    x1: float = trial.suggest_float("x1", 0.1, 3)
+    x2: float = trial.suggest_float("x2", 0.1, 3, log=True)
+    x3: int = trial.suggest_int("x3", 2, 4, log=True)
+    x4 = trial.suggest_categorical("x4", [0.1, 1.0, 10.0])
+    assert isinstance(x4, float)
+    return x1 + x2 * x3 + x4
```
[the remark this answers is not in the document]
Comment (contramundum53) [reply tests/integration_tests/test_shap.py:27]: I think that's a good idea.

File: optuna/importance/_mean_decrease_impurity.py [hunk 6]
```diff
@@ -71,42 +74,25 @@ def evaluate(
...
         feature_importances = forest.feature_importances_
-        feature_importances_reduced = numpy.zeros(len(distributions))
-        numpy.add.at(feature_importances_reduced, encoded_column_to_column, feature_importances)
 
-        param_importances = OrderedDict()
-        self._param_names = list(distributions.keys())
-        for i in feature_importances_reduced.argsort()[::-1]:
-            param_importances[self._param_names[i]] = feature_importances_reduced[i].item()
+        # Untransform feature importances to param importances
+        #  by adding up relevant feature importances
```
[the remark this answers is not in the document]
Comment (contramundum53) [reply optuna/importance/_mean_decrease_impurity.py]: Thanks. I applied the suggestion.

File: optuna/importance/_fanova/_evaluator.py (hunk 1 above)
Comment (c-bata) [reply optuna/importance/_fanova/_evaluator.py:107]: It looks that this suggestion is still not applied :eyes:

**Review by c-bata — COMMENTED** [review]
Looks almost good to me!

**Review by nzw0301 — APPROVED** [review]
LGTM. Thanks!

**Review by c-bata — APPROVED** [review]
LGTM

**Pull request merged**